pkg/api/component_report.go

```diff
@@ -0,0 +1,628 @@
+)
+
+func GetComponentReportFromBigQuery(client *bigquery.Client,
+	baseRelease, sampleRelease, component, capability, platform, upgrade, arch, network, testId, groupBy string,
+	baseStartTime, baseEndTime, sampleStartTime, sampleEndTime time.Time) (apitype.ComponentReport, []error) {
+	generator := componentReportGenerator{
+		client:          client,
+		baseRelease:     baseRelease,
+		sampleRelease:   sampleRelease,
+		component:       component,
+		capability:      capability,
+		platform:        platform,
+		upgrade:         upgrade,
+		arch:            arch,
+		network:         network,
+		testId:          testId,
+		groupBy:         groupBy,
+		baseStartTime:   baseStartTime,
+		baseEndTime:     baseEndTime,
+		sampleStartTime: sampleStartTime,
+		sampleEndTime:   sampleEndTime,
+		confidence:      95,
+		minimumFailure:  3,
+		pityFactor:      5,
+	}
+	return generator.GenerateReport()
+}
+
+type componentReportGenerator struct {
+	client          *bigquery.Client
+	baseRelease     string
+	sampleRelease   string
+	component       string
+	capability      string
+	platform        string
+	upgrade         string
+	arch            string
+	network         string
+	testId          string
+	groupBy         string
+	baseStartTime   time.Time
+	baseEndTime     time.Time
+	sampleStartTime time.Time
+	sampleEndTime   time.Time
+	ignoreMissing   bool
+	minimumFailure  int
+	confidence      int
+	pityFactor      int
+}
+
+func (c *componentReportGenerator) GenerateReport() (apitype.ComponentReport, []error) {
+	baseStatus, sampleStatus, errs := c.getTestStatusFromBigQuery()
+	report := c.generateComponentTestReport(baseStatus, sampleStatus)
+	return report, errs
+}
+
+func (c *componentReportGenerator) getTestStatusFromBigQuery() (
+	map[apitype.ComponentTestIdentification]apitype.ComponentTestStats,
+	map[apitype.ComponentTestIdentification]apitype.ComponentTestStats,
+	[]error,
+) {
+	errs := []error{}
+	// NOTE: casting a couple datetime columns to timestamps, it does appear they go in as UTC, and thus come out
+	// as the default UTC correctly.
+	// Annotations and labels can be queried here if we need them.
+	queryString := `SELECT
+			network,
+			upgrade,
+			arch,
+			platform,
+			test_id,
+            ANY_VALUE(test_name) AS test_name,
+            COUNT(test_id) AS total_count,
+			SUM(success_val) AS success_count,
+			SUM(flake_count) AS flake_count ` +
+		"FROM `ci_analysis_us.junit` " +
+		`WHERE TIMESTAMP(modified_time) >= @From AND TIMESTAMP(modified_time) < @To `
+
+	if c.upgrade != "" {
+		queryString = queryString + ` AND upgrade = "` + c.upgrade + `"`
+	}
+	if c.arch != "" {
+		queryString = queryString + ` AND arch = "` + c.arch + `"`
+	}
+	if c.network != "" {
+		queryString = queryString + ` AND network = "` + c.network + `"`
+	}
+	if c.platform != "" {
+		queryString = queryString + ` AND platform = "` + c.platform + `"`
+	}
+	if c.testId != "" {
+		queryString = queryString + ` AND test_id = "` + c.testId + `"`
+	}
```

Comment:
This opens you to SQL injection since these probably end up eventually come from user input on the API.  I'd suggest using QueryParameter which will avoid that kind of thing.  https://pkg.go.dev/cloud.google.com/go/bigquery#QueryParameter

There's examples in ci-tools

https://github.com/openshift/ci-tools/blob/master/pkg/jobrunaggregator/jobrunaggregatorlib/ci_data_client.go#L262